### PR TITLE
[FEATURE] Partial subtitles support

### LINKED ISSUE #79: [FEATURE] Add support for subtitles.
I have seen the subtitles are available in most the providers via the settings option in the video player but the mov-cli doesn't extract them.

will it be supported? or possible external support to be included like subscene or others?

### PR DESCRIPTION
**Pull request by shinysocks**
implements: #79 
* created a Subtitles class which contains methods get_movie_subs() and get_tv_subs() which query api.opensubtitles.com and return a link to subtitle files for the respective movie or episode.
* commented the VLC and MPV classes with --sub-file flags to show how subtitles can be added to those players.

### CODE REVIEW

**Comment by shinysocks** [comment]
[api.opensubtitles.com docs](https://opensubtitles.stoplight.io/docs/opensubtitles-api/573f76acc1493-subtitle)

**Comment by shinysocks** [comment]
check on discord, this is not integrated!!

File: mov_cli/subtitles.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import requests
+
+
+class Subtitles:
+    """
+    pull subtitles from api.opensubtitles.com and return
+    a link to the .srt file which can be passed into vlc or mpv
+    """
+
+    def __init__(self, key, language="en"):
```
Comment (THEGOLDENPRO) [inline mov_cli/subtitles.py]: Can `key` be optional? E.g. be `None` if the user provides no API key.

**Comment by THEGOLDENPRO** [comment]
> check on discord, this is not integrated!!

ah, my bad still a draft. I'm going to make some corrections however.

File: mov_cli/subtitles.py (hunk 1 above)
Comment (shinysocks) [reply mov_cli/subtitles.py]: not really, the new opensubtitles api forces keys.. no anonymous downloads! Still, we could create an account for an organization but it still has limits.

**Comment by shinysocks** [comment]
sounds good, otherwise lemme know if there's anything I can change on my end

File: mov_cli/subtitles.py (hunk 1 above)
Comment (shinysocks) [reply mov_cli/subtitles.py]: also I was thinking the key would be inputted into the config file, not sure how to do that

Comment (THEGOLDENPRO) [reply mov_cli/subtitles.py]: > not really, the new opensubtitles api forces keys.. no anonymous downloads! Still, we could create an account for an organization but it still has limits.

ah okay

Comment (THEGOLDENPRO) [reply mov_cli/subtitles.py]: > also I was thinking the key would be inputted into the config file, not sure how to do that

I can do that, we just need to add it as a field in the config class and then pass the config class as a parameter to the subtitles class.

**Comment by shinysocks** [comment]
thanks for your changes! any ideas on how to set it up with the rest of the codebase?

**Comment by THEGOLDENPRO** [comment]
> thanks for your changes! any ideas on how to set it up with the rest of the codebase?

yeah, I'm going to look into it. :+1: First I need to get an API key and test this.

**Comment by THEGOLDENPRO** [comment]
how is the password not the same when I copy and pasted it from my password manager :sob: WTF

Screenshot_20240315_205822

**Comment by THEGOLDENPRO** [comment]
I made a little test suite for this. You can run it with this command from the root directory:
```sh
python scripts/test_subtitles.py
```

**Comment by THEGOLDENPRO** [comment]
I'll work more on this another day. Thanks for your contributions.

**Comment by shinysocks** [comment]
looks awesome, i'd love to contribute more

**Comment by THEGOLDENPRO** [comment]
@Poseidon444 @r3tr0ananas Also can we remove Codacy, it's bugging me lmao.
image

I think if we want proper linting and type checking we should use something like ruff (which we already have). It's a hell of a lot faster and more accurate and if you want ruff to be more strict feel free to modify its [toml file](https://github.com/mov-cli/mov-cli/blob/v4/ruff.toml). Up to you guys are the end of the day though.

**Comment by THEGOLDENPRO** [comment]
> looks awesome, i'd love to contribute more

feel free 🙏

**Comment by ananasmoe** [comment]
> @Poseidon444 @r3tr0ananas Also can we remove Codacy, it's bugging me lmao. 

Yeah, we can.

**Comment by shinysocks** [comment]
I still got time today to get this working if you wanna explain where I should call the subtitle methods to integrate it, I just don't really understand the way mov-cli works..

**Comment by THEGOLDENPRO** [comment]
mov-cli v4 is plugin based meaning most likely the subtitles class will be called from the scraper in the plugin itself rather than mov-cli although I am still uncertain.

I'm not too good at explaining things but I'll try.

Here's a brief path of how mov-cli operates with plugins:
```sh
mov-cli -s films {query}
```
Results in these operations:
```
cli module --> plugin ("films" in this case) --> plugin's scraper --> Scraper.search(query) --> prompt results to user --> Scraper.scrape_metadata_episdoes(result) --> prompt episodes to user --> Scraper.scrape(result, episode) --> Player.play(media)
```

I'm still trying to come up with a good way to implement subtitles, I think there will be a bunch of changes that will have to be done with the codebase before subtitles can be implemented.

**Comment by shinysocks** [comment]
okay thank you for the explaination, I'll try to use the template stuff that @EbanNox sent in discord and put the subtitle functionality into some sort of plugin in the mean time

**Comment by THEGOLDENPRO** [comment]
Just gonna dump my current thoughts here publically so I don't end up forgetting.

It should be the plugin's job to retrieve subtitles and pass it to the `Media` class, as if we do it natively (outside of the plugin's scraper) we will only have the user's query to work with in regards of determining which subtitle is actually the correct one for that media. The scraper can do a better job at this as it has the entire source to work with.

The subtitles url returned by the `Subtitles` class can be passed to the `subtitles` parameter in the `Media` class that the player will then handle. I say "handle" because with MPV you can pass the url as a subtitles file no problem and it just works but for VLC you will have to download the subs first then pass it; that can be done very easily from the VLC player class however.

Okay but how about if you don't want subtitles for the media you watching? This is where we must introduce a cli option like ``--subs`` or ``--subtitles``. These will be boolean flags that you set whenever you want to pull subtitles. Although now we have another issue, remember that we gave the scraper control over subtitles. Now we'll need to tell the scraper when to pull subtitles and when to not. 

To do this we can introduce a new parameter to `Scraper.scrape()` like ``scrape_subtitles``:
```python
def scrape(self, metadata: Metadata, episode: Optional[EpisodeSelector] = None, scrape_subtitles: bool = False) -> Series | Movie:
    ...
```

Togther with the command options ``--subs`` and ``--subtitles`` we must also introduce a [toml key](https://github.com/mov-cli/mov-cli/wiki/Configuration#toml-keys--cli-options) where if the `default` key is set to true subtitles will be pulled by default, something like so:
```toml
[mov-cli.subtitles]
default = false
open_subtitles_key = "{your-key}"
```

**Comment by THEGOLDENPRO** [comment]
> Now we'll need to tell the scraper when to pull subtitles and when to not.
> 
> To do this we can introduce a new parameter to `Scraper.scrape()` like `scrape_subtitles`:
> 
> ```python
> def scrape(self, metadata: Metadata, episode: Optional[EpisodeSelector] = None, scrape_subtitles: bool = False) -> Series | Movie:
>     ...
> ```

🎉 This pull request is what is going to make this possible: #261

**Comment by EbanNox** [comment]
> Just gonna dump my current thoughts here publically so I don't end up forgetting.
> 
> It should be the plugin's job to retrieve subtitles and pass it to the `Media` class, as if we do it natively (outside of the plugin's scraper) we will only have the user's query to work with in regards of determining which subtitle is actually the correct one for that media. The scraper can do a better job at this as it has the entire source to work with.
> 
> The subtitles url returned by the `Subtitles` class can be passed to the `subtitles` parameter in the `Media` class that the player will then handle. I say "handle" because with MPV you can pass the url as a subtitles file no problem and it just works but for VLC you will have to download the subs first then pass it; that can be done very easily from the VLC player class however.
> 
> Okay but how about if you don't want subtitles for the media you watching? This is where we must introduce a cli option like ``--subs`` or ``--subtitles``. These will be boolean flags that you set whenever you want to pull subtitles. Although now we have another issue, remember that we gave the scraper control over subtitles. Now we'll need to tell the scraper when to pull subtitles and when to not. 
> 
> To do this we can introduce a new parameter to `Scraper.scrape()` like ``scrape_subtitles``:
> ```python
> def scrape(self, metadata: Metadata, episode: Optional[EpisodeSelector] = None, scrape_subtitles: bool = False) -> Series | Movie:
>     ...
> ```
> 
> Togther with the command options ``--subs`` and ``--subtitles`` we must also introduce a [toml key](https://github.com/mov-cli/mov-cli/wiki/Configuration#toml-keys--cli-options) where if the `default` key is set to true subtitles will be pulled by default, something like so:
> ```toml
> [mov-cli.subtitles]
> default = false
> open_subtitles_key = "{your-key}"
> ```

I my self think of it like this... have the sub plugin behave like a scraper that just passes on to another plugin scraper if called...

Like:

Mov-cli -s sub.films.vadapav "eng" "spider man" -p mpv etc....
[Mov-cli--> calls sub scraper(fetches urls and then calls vadapav scraper)--> vadapav(fetches media and passes on to player methos) --> play.media(having the sub and media mixed together and end up on the player)

But as this is having the -s option used in the example, I think the same method can be devised when iplmlementing plugin passing new arg options to mov-cli it self. I'm just gonna leave this here as a mind map for the end result .

**Comment by THEGOLDENPRO** [comment]
I don't think this subtitles pr will do. Sadly if you don't get subs directly from the providers you will either have no subtitles or incorrect and delayed subtitles. Open Subtitles also has heavy rate limits for non developers so effecting basically every user.

I don't think this is the implementation I want to encourage in mov-cli plugins as it will lead to the problems I listed above.

Subtitles should be grabbed directly from the media provider itself, just like how the [vadapav scraper does it](https://github.com/JDALab/mov-cli-films/blob/49c4750b653f21aa0e800b340a31cc45e446d55d/mov_cli_films/vadapav/scraper.py#L147).

I'll merge a few things from this pr though like the player subtitle args. Thanks @shinysocks for your contribution.

**Pull request merged**